<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ChIP and ATAC-seq data processing pipeline • chompR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="ChIP and ATAC-seq data processing pipeline">
<meta property="og:description" content="An R-based pipeline to analyse chromatin modification high throughput 
  sequencing data. This pipeline performs quality control of reads, performs peak-calling, 
  differential peak-calling, and pathway enrichment analyses.">
<meta property="og:image" content="https://anilchalisey.github.io/chompR/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">chompR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/anilchalisey/chompR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<div id="chompr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#chompr" class="anchor"></a>chompR <img src="reference/figures/logo.png" align="right">
</h1></div>
<p>The chompR package is designed to perform peak-calling and differential analysis of ChIP or ATAC-seq data. It provides tools to:</p>
<ul>
<li>perform quality control of raw reads in fastq format and generate HTML reports (using FASTQC and MULTIQC)</li>
<li>perform read alignment using HISAT2 and sort and filter alignment files</li>
<li>perform quality control of aligned reads</li>
<li>perform peak-calling using MACS2</li>
<li>run differential analysis using DESeq2, and to export the results into easily readable tab-delimited files</li>
<li>perform pathway analysis</li>
</ul>
<p>The chompR package does not attempt to replace the tools it uses, but instead provides an automated environment in which to use them.</p>
<p>The website for chompR may be found <a href="https://anilchalisey.github.io/chompR/">here</a>.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<div id="command-line-tools" class="section level2">
<h2 class="hasAnchor">
<a href="#command-line-tools" class="anchor"></a>Command line tools</h2>
<p>The workflow used by chompR requires the installation of several command-line tools: <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Fastqc%5D">Fastqc (v0.11.7)</a>, <a href="http://multiqc.info/">MultiQC (v1.6)</a>, <a href="https://github.com/OpenGene/fastp">fastp (v0.19.4)</a>, <a href="http://www.htslib.org/">SAMtools (v1.9)</a>, <a href="http://lomereiter.github.io/sambamba/">Sambamba (v0.6.7)</a>, <a href="https://ccb.jhu.edu/software/hisat2/index.shtml">HISAT2 (v2.1.0)</a>, <a href="https://github.com/taoliu/MACS">MACS (v2.1.0)</a>, and <a href="http://subread.sourceforge.net/">featureCounts (v1.6.2)</a>. Installation instructions for these may be found at their respective websites, but a guide is given below for convenience.</p>
<p>If using a Unix-based system, open up a terminal and follow the commands as is. If using the windows subsystem for linux (WSL) on Windows 10 then WSL must first be set up as detailed <a href="https://docs.microsoft.com/en-us/windows/wsl/install-win10">here</a>. Once WSL is up and running, then the tools may be installed as on any Unix-based system.</p>
<p>It is assumed that root priviliges are available (if necessary) - if not, then a system administrator may need to install these for you.</p>
<div id="fastqc" class="section level3">
<h3 class="hasAnchor">
<a href="#fastqc" class="anchor"></a>Fastqc</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="fu">wget</span> https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.7.zip</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="fu">unzip</span> fastqc_v0.11.7.zip</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="bu">cd</span> FastQC</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="fu">chmod</span> 755 fastqc</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co"># If using WSL</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="bu">cd</span> /usr/local/bin</a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="fu">sudo</span> ln -s ~/Fastqc/fastqc .</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co"># If using other Unix</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="bu">cd</span> ~/bin</a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="fu">ln</span> -s ~/Fastqc/fastqc .</a></code></pre></div>
</div>
<div id="multiqc" class="section level3">
<h3 class="hasAnchor">
<a href="#multiqc" class="anchor"></a>MultiQC</h3>
<p>To run MultiQC ensure there is a working python distribution.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="ex">pip</span> install multiqc</a></code></pre></div>
</div>
<div id="fastp" class="section level3">
<h3 class="hasAnchor">
<a href="#fastp" class="anchor"></a>fastp</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="fu">wget</span> http://opengene.org/fastp/fastp</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="fu">chmod</span> 755 fastp</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co"># If using WSL</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="bu">cd</span> /usr/local/bin</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="fu">sudo</span> ln -s ~/fastp .</a>
<a class="sourceLine" id="cb3-8" data-line-number="8"></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co"># If using other Unix</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="bu">cd</span> ~/bin</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="fu">ln</span> -s ~/fastp .</a></code></pre></div>
</div>
<div id="sambamba" class="section level3">
<h3 class="hasAnchor">
<a href="#sambamba" class="anchor"></a>Sambamba</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="fu">wget</span> https://github.com/biod/sambamba/releases/download/v0.6.7/sambamba_v0.6.7_linux.tar.bz2</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="fu">tar</span> xvjf sambamba_v0.6.7_linux.tar.bz2</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="fu">chmod</span> 755 sambamba</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co"># If using WSL</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="bu">cd</span> /usr/local/bin</a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="fu">sudo</span> ln -s ~/sambamba .</a>
<a class="sourceLine" id="cb4-9" data-line-number="9"></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co"># If using other Unix</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="bu">cd</span> ~/bin</a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="fu">ln</span> -s ~/sambamba .</a></code></pre></div>
</div>
<div id="hisat2" class="section level3">
<h3 class="hasAnchor">
<a href="#hisat2" class="anchor"></a>Hisat2</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="fu">wget</span> ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/downloads/hisat2-2.1.0-Linux_x86_64.zip</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="fu">unzip</span> hisat2-2.1.0-Linux_x86_64.zip</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co"># If using WSL</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="bu">cd</span> /usr/local/bin</a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="fu">sudo</span> ln -s ~/hisat2-2.1.0/hisat2_* .</a>
<a class="sourceLine" id="cb5-8" data-line-number="8"></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co"># If using other Unix</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="bu">cd</span> ~/bin</a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="fu">ln</span> -s ~/hisat2-2.1.0/hisat2_* .</a></code></pre></div>
</div>
<div id="macs2" class="section level3">
<h3 class="hasAnchor">
<a href="#macs2" class="anchor"></a>MACS2</h3>
<p>To run MACS2 ensure there is a working python (&gt;2.7) distribution. GCC is also required to compile <code>.c</code> codes (this should be pre-installed with your UNIX distribution) and python header files are needed. Therefore, <code>python-dev</code> must also be installed.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="fu">sudo</span> apt install python-dev</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="ex">pip</span> install MACS2</a></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="fu">wget</span> -O subread.tar.gz https://sourceforge.net/projects/subread/files/subread-1.6.2/subread-1.6.2-Linux-x86_64.tar.gz/download</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="fu">tar</span> xvf subread.tar.gz</a>
<a class="sourceLine" id="cb7-4" data-line-number="4"></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="co"># If using WSL</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="bu">cd</span> /usr/local/bin</a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="fu">sudo</span> ln -s ~/subread-1.6.2-Linux-x86_64/bin/sub* .</a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="fu">sudo</span> ln -s ~/subread-1.6.2-Linux-x86_64/bin/exactSNP .</a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="fu">sudo</span> ln -s ~/subread-1.6.2-Linux-x86_64/bin/featureCounts .</a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="fu">sudo</span> ln -s ~/subread-1.6.2-Linux-x86_64/bin/utilities/* .</a>
<a class="sourceLine" id="cb7-11" data-line-number="11"></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"><span class="co"># If using other Unix</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"><span class="bu">cd</span> ~/bin</a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="fu">sudo</span> ln -s ~/subread-1.6.2-Linux-x86_64/bin/sub* .</a>
<a class="sourceLine" id="cb7-15" data-line-number="15"><span class="fu">sudo</span> ln -s ~/subread-1.6.2-Linux-x86_64/bin/exactSNP .</a>
<a class="sourceLine" id="cb7-16" data-line-number="16"><span class="fu">sudo</span> ln -s ~/subread-1.6.2-Linux-x86_64/bin/featureCounts .</a>
<a class="sourceLine" id="cb7-17" data-line-number="17"><span class="fu">sudo</span> ln -s ~/subread-1.6.2-Linux-x86_64/bin/utilities/* .</a></code></pre></div>
</div>
<div id="r-package-dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#r-package-dependencies" class="anchor"></a>R package dependencies</h3>
<p>There are a number of dependencies to the chompR package detailed in the <code>DESCRIPTION</code> file. These are either CRAN or Bioconductor packages. These dependencies will automatically install when installing chompR.</p>
</div>
<div id="install-chompr" class="section level3">
<h3 class="hasAnchor">
<a href="#install-chompr" class="anchor"></a>Install chompR</h3>
<p>Once all dependencies are installed, then chompR may be installed as follows:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"anilchalisey/chompR"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>The entire workflow of chompR may be run by a call to a single function <code><a href="reference/run_chip.html">run_chip()</a></code> or <code><a href="reference/run_atac.html">run_atac()</a></code>, as shown below.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">result_chip &lt;-<span class="st"> </span><span class="kw"><a href="reference/run_chip.html">run_chip</a></span>(<span class="dt">sample.info =</span> <span class="st">`</span><span class="dt">system.file("extdata", "paired-example.txt", pkg ="chompR")</span><span class="st">`</span>,</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">                  <span class="dt">reference =</span> <span class="kw">c</span>(<span class="st">"WT"</span>, <span class="st">"KO1"</span>, <span class="st">"KO2"</span>), <span class="dt">species =</span> <span class="st">"human"</span>, <span class="dt">output.dir =</span> <span class="st">"results-chip"</span>,</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">                  <span class="dt">threads =</span> <span class="ot">NULL</span>, <span class="dt">index.dir =</span> <span class="ot">NULL</span>)</a></code></pre></div>
<p>The arguments to the <code><a href="reference/run_chip.html">run_chip()</a></code> and <code><a href="reference/run_atac.html">run_atac()</a></code> functions are:</p>
<div id="sample-info" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-info" class="anchor"></a>sample.info:</h3>
<p>This is the path to a tab-delimited file with at least the columns:</p>
<ul>
<li>
<code>condition</code>: treatment or condition labels on the basis of which the differential analysis will be performed (e.g. “CONTROL”, “TREATED”; or “WILD-TYPE”, “KNOCK-OUT1”, “KNOCK-OUT2”).</li>
<li>
<code>sample</code>: names of the samples - these labels will be used in plots and as sample column headers in the output.</li>
<li>
<code>file1</code>: absolute or relative path to fastq files.</li>
</ul>
<p>Optional columns include:</p>
<ul>
<li>
<code>file2</code>: if fastq files and PE reads, then this column should also be present, specifying the absolute or relative paths to the second pair of fastq reads.<br>
# <code>batch</code>: if a batch effect is to be included in the design, then this should be identified under this column (e.g. litter number or sequencing run). # <code>input1</code>: if ChIP-seq and input samples were used then the path to the fastq file for the input matching file1 in the same row should be included here. # <code>input2</code>: same as input1, but if PE reads, then the second pair should be specified here.</li>
</ul>
<p>Dummy examples of the tab-delimited files accepted by chompR may be found at <code>system.file("extdata", "paired-example.txt", pkg = "chompR")</code>, <code>system.file("extdata", "sinle-example.txt", pkg = "chompR")</code>, and <code>system.file("extdata", "quants-example.txt", pkg = "chompR")</code>.</p>
</div>
<div id="reference" class="section level3">
<h3 class="hasAnchor">
<a href="#reference" class="anchor"></a>reference:</h3>
<p>The order in which the condition labels should be evaluated when performing differential analysis. For example, c(“A”, “B”, “C”, “D”) would mean “A” is the reference condition to which “B”, “C” and “D” are compared; in addition, “C” and “D” will be compared to “B”, and “D” will be compared to “C”. If  then the comparisons will be arranged alphabetically.</p>
</div>
<div id="species" class="section level3">
<h3 class="hasAnchor">
<a href="#species" class="anchor"></a>species:</h3>
<p>May be one of “human” or “mouse” - other options will be added in later versions.</p>
</div>
<div id="output-dir" class="section level3">
<h3 class="hasAnchor">
<a href="#output-dir" class="anchor"></a>output.dir:</h3>
<p>Directory to which results should be saved.</p>
</div>
<div id="threads" class="section level3">
<h3 class="hasAnchor">
<a href="#threads" class="anchor"></a>threads:</h3>
<p>Number of threads to be used when parallelisation is possible. If <code>NULL</code> then one less than the maximum numbre of threads available will be used.</p>
</div>
<div id="index-dir" class="section level3">
<h3 class="hasAnchor">
<a href="#index-dir" class="anchor"></a>index.dir:</h3>
<p>The path to the hisat2 index.</p>
<p>Other arguments are also possible, and details for these may be found in the manual, but for most users, the default settings are satisfactory.</p>
<p>Importantly, if the user has already run the analysis once and read QC and alignment has already occurred, then this will be detected and these steps will be skipped (provided the same output directory is specified).</p>
</div>
</div>
<div id="other-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#other-functions" class="anchor"></a>Other functions</h2>
<p>The package also contains several functions to directly run several tools from within R. These all begin with  and include , , , , , , , , . Additional functions include those to import MACS2 format peak files into R () and to annotate peaks ().</p>
</div>
<div id="output" class="section level2">
<h2 class="hasAnchor">
<a href="#output" class="anchor"></a>Output</h2>
<p>Once complete, all the results will be saved in the specified output directory with the following structure:</p>
<pre><code>#&gt;                  levelName
#&gt; 1  results                
#&gt; 2   ¦--alignment_stats.txt
#&gt; 3   ¦--bam                
#&gt; 4   ¦--consensus          
#&gt; 5   ¦--cross-correlation  
#&gt; 6   ¦--differential       
#&gt; 7   ¦--fastqc             
#&gt; 8   ¦--index              
#&gt; 9   ¦--multiqc            
#&gt; 10  ¦--filteredbam        
#&gt; 11  ¦--shifted            
#&gt; 12  ¦--trimmed            
#&gt; 13  ¦--peaks              
#&gt; 14  ¦--DESeq2             
#&gt; 15  °--fragmentLength</code></pre>
<p>The directories <code>DESeq2</code>, <code>fastqc</code>, and <code>multiqc</code> contain the results of the respective analyses including relevant plots. The <code>index</code> directory contains the hisat2 index for human or mouse; it will only be creeated if a pre-existing index was not specified beforehand. The <code>bam</code>, <code>filteredbam</code>, <code>trimmedbam</code>, and <code>shifted</code> directories contain BAM files. The <code>shifted</code> directory only exists for ATAC-seq analysis and contains BAM files whose reads have been shifted to take account of transposon insertion. The <code>peaks</code> and <code>consensus</code> directories contain MACS2 called peaks and consensus peaks respectively. The <code>differential</code> directory contains the results of the a binary differential analysis i.e. peaks present or absent in each sample. The <code>cross-correlation</code> and <code>fragmentlength</code> directories are only created for ChIP_seq or ATAC-seq respectively (i.e. they are mutually exclusive) and contain QC plots following alignment. The <code>alignment_stats.txt</code> file contains simple alignment statistics.</p>
</div>
<div id="running-only-part-of-the-chompr-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#running-only-part-of-the-chompr-pipeline" class="anchor"></a>Running only part of the chompR pipeline</h2>
<p>Importantly, if the user has already run the analysis once and read QC and alignment has already occurred, then this will be detected and these steps will be skipped (provided the same output directory is specified).</p>
</div>
<div id="pathway-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#pathway-analysis" class="anchor"></a>Pathway analysis</h2>
<p>GREAT analysis (using the GREAT website) is performed using the package rGREAT. A wrapper script, which converts regions into hg19 co-ordinates (if using “human”) and then runs the rGREAT package, is included.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">peaks &lt;-<span class="st"> </span><span class="kw"><a href="reference/peak2Granges.html">peak2Granges</a></span>(<span class="st">"peaks/results.narrowPeak"</span>)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">pathway &lt;-<span class="st"> </span><span class="kw"><a href="reference/run_great.html">run_great</a></span>(<span class="dt">regions =</span> peaks, <span class="dt">species =</span> <span class="st">"human"</span>)</a></code></pre></div>
<p>This may be repeated as necessary for the other comparisons.</p>
</div>
<div id="about-chompr" class="section level2">
<h2 class="hasAnchor">
<a href="#about-chompr" class="anchor"></a>About chompR</h2>
<p>The chompR package has been developed in the Chris O’Callaghan Group at the Centre for Cellular and Molecular Biology, University of Oxford by Anil Chalisey and Chris O’Callaghan.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/anilchalisey/chompR">https://​github.com/​anilchalisey/​chompR</a>
</li>
<li>Report a bug at <br><a href="https://github.com/anilchalisey/chompR/issues">https://​github.com/​anilchalisey/​chompR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/anilchalisey">Anil Chalisey</a> <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/anilchalisey">Anil Chalisey</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
