<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Wrapper script to run HISAT2. — run_hisat2 • chompR</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Wrapper script to run HISAT2. — run_hisat2" />

<meta property="og:description" content="Script to align reads to a reference genome using hisat2.  This
  requires an existing index which may be created using hisat2 itself.
  Commonly used genome indices may also be downloaded from the HISAT2
  homepage." />

<meta property="og:image" content="https://anilchalisey.github.io/chompR/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">chompR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/anilchalisey/chompR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper script to run HISAT2.</h1>
    <small class="dont-index">Source: <a href='https://github.com/anilchalisey/chompR/blob/master/R/run_hisat2.R'><code>R/run_hisat2.R</code></a></small>
    <div class="hidden name"><code>run_hisat2.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Script to align reads to a reference genome using hisat2.  This
  requires an existing index which may be created using hisat2 itself.
  Commonly used genome indices may also be downloaded from the HISAT2
  homepage.</p>
    
    </div>

    <pre class="usage"><span class='fu'>run_hisat2</span>(<span class='kw'>hisat2</span> <span class='kw'>=</span> <span class='st'>"hisat2"</span>, <span class='kw'>idx</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>mate1</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>mate2</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>fastq</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>fasta</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>softClipPenalty</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>noSoftClip</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>noSplice</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>knownSplice</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>strand</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>tmo</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>maxAlign</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>secondary</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>minInsert</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>maxInsert</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>nomixed</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>nodiscordant</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>rgid</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>non_deterministic</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>hisat2</th>
      <td><p>Path to hisat2 (if using WSL, then this should be the
full path on the linux subsystem)</p></td>
    </tr>
    <tr>
      <th>idx</th>
      <td><p>The basename of the index for the reference genome. The basename
is the name of any of the index files up to but not including the final
.1.ht2, etc.</p></td>
    </tr>
    <tr>
      <th>mate1</th>
      <td><p>Comma-separated list of files containing mate 1s (filename
usually includes _1)</p></td>
    </tr>
    <tr>
      <th>mate2</th>
      <td><p>Comma-separated list of files containing mate 2s (filename
usually includes _2). Sequences specified with this option must correspond
file-for-file and read-for-read with those specified in .</p></td>
    </tr>
    <tr>
      <th>fastq</th>
      <td><p>Logical indicating if reads are FASTQ files.</p></td>
    </tr>
    <tr>
      <th>fasta</th>
      <td><p>Logical indicating if reads are FASTA files.</p></td>
    </tr>
    <tr>
      <th>softClipPenalty</th>
      <td><p>Sets the maximum (MX) and minimum (MN) penalties for
soft-clipping per base, both integers.  Must be given in the format "MX,MN".</p></td>
    </tr>
    <tr>
      <th>noSoftClip</th>
      <td><p>Logical indicating whether to disallow soft-clipping.</p></td>
    </tr>
    <tr>
      <th>noSplice</th>
      <td><p>Logical indicating whether to switch off spliced alignment,
e.g., for DNA-seq analysis.</p></td>
    </tr>
    <tr>
      <th>knownSplice</th>
      <td><p>Path to text file containing known splice sites.</p></td>
    </tr>
    <tr>
      <th>strand</th>
      <td><p>Specify strand-specific information.  Default is unstranded.</p></td>
    </tr>
    <tr>
      <th>tmo</th>
      <td><p>Logical indicating whether to report only those reads aligning to
known transcripts.</p></td>
    </tr>
    <tr>
      <th>maxAlign</th>
      <td><p>Integer indicating the maximum number of distinct primary
alignments to search for each read.</p></td>
    </tr>
    <tr>
      <th>secondary</th>
      <td><p>Logical indicating whether to report secondary alignments.</p></td>
    </tr>
    <tr>
      <th>minInsert</th>
      <td><p>The minimum fragment length for valid paired-end alignments.
This option is valid only with noSplice = TRUE.</p></td>
    </tr>
    <tr>
      <th>maxInsert</th>
      <td><p>The maximum fragment length for valid paired-end alignments.
This option is valid only with noSplice = TRUE.</p></td>
    </tr>
    <tr>
      <th>nomixed</th>
      <td><p>By default, when hisat2 cannot find a concordant or discordant
alignment for a pair, it then tries to find alignments for the individual
mates. If TRUE, this option disables that behavior.</p></td>
    </tr>
    <tr>
      <th>nodiscordant</th>
      <td><p>By default, hisat2 looks for discordant alignments if it
cannot find any concordant alignments.  If true, this option disables that
behavior.</p></td>
    </tr>
    <tr>
      <th>threads</th>
      <td><p>an integer value indicating the number of workers to be used. If NULL then one less than the
maximum number of cores will be used. [DEFAULT = NULL].</p></td>
    </tr>
    <tr>
      <th>rgid</th>
      <td><p>Character string, to which the read group ID is set.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>If TRUE, print nothing except alignments and serious errors.</p></td>
    </tr>
    <tr>
      <th>non_deterministic</th>
      <td><p>When set to TRUE, HISAT2 re-initializes its
pseudo-random generator for each read using the current time.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Alignment file in SAM format</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='fu'>run_hisat2</span>(<span class='kw'>hisat2</span> <span class='kw'>=</span> <span class='st'>"hisat2"</span>, <span class='kw'>idx</span> <span class='kw'>=</span> <span class='st'>"../prana/data-raw/index/UCSC.hg19"</span>,
<span class='kw'>mate1</span> <span class='kw'>=</span> <span class='st'>"../prana/data-raw/seqFiles/HB1_sample_1.fastq.gz"</span>,
<span class='kw'>mate2</span> <span class='kw'>=</span> <span class='st'>"../prana/data-raw/seqFiles/HB1_sample_2.fastq.gz"</span>,
<span class='kw'>fastq</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>fasta</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>softClipPenalty</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>noSoftClip</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
<span class='kw'>noSplice</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>knownSplice</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>strand</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>tmo</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
<span class='kw'>maxAlign</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>secondary</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>minInsert</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>maxInsert</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>nomixed</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>nodiscordant</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
<span class='kw'>threads</span> <span class='kw'>=</span> (<span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/parallel/topics/detectCores'>detectCores</a></span>() - <span class='fl'>1</span>), <span class='kw'>rgid</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
<span class='kw'>non_deterministic</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://github.com/anilchalisey'>Anil Chalisey</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

